app/views/handlers/confirmation_email.py

```diff
@@ -34,17 +60,59 @@ def get_page_title(self):
             return gettext("Error: {page_title}").format(page_title=self.page_title)
         return self.page_title
 
-    @staticmethod
-    def handle_post(session_store: SessionStore):
-        session_store.session_data.confirmation_email_count += 1
-        session_store.save()
+    def _publish_fulfilment_request(self):
+        topic_id = current_app.config["EQ_SUBMISSION_CONFIRMATION_TOPIC_ID"]
+        fulfilment_request = ConfirmationEmailFulfilmentRequest(
+            self.form.email.data, self._session_store.session_data, self._schema
+        )
+        try:
+            return current_app.eq["publisher"].publish(
+                topic_id, message=fulfilment_request.payload
+            )
+        except PublicationFailed:
+            raise ConfirmationEmailFulfilmentRequestPublicationFailed
 
-    @staticmethod
-    def is_limit_reached():
-        session_store = get_session_store()
+    def handle_post(self):
+        self._publish_fulfilment_request()
+        self._session_store.session_data.confirmation_email_count += 1
+        self._session_store.save()
 
+    @staticmethod
+    def is_limit_reached(session_data: SessionData) -> bool:
         return (
-            session_store
-            and session_store.session_data.confirmation_email_count
+            session_data.confirmation_email_count
             >= current_app.config["CONFIRMATION_EMAIL_LIMIT"]
         )
+
+
+@dataclass
+class ConfirmationEmailFulfilmentRequest(FulfilmentRequest):
+    email_address: str
+    session_data: SessionData
+    schema: QuestionnaireSchema
+
+    @property
+    def payload(self) -> bytes:
+        message = {
+            "event": {
+                "type": "FULFILMENT_REQUESTED",
+                "source": "QUESTIONNAIRE_RUNNER",
+                "channel": "EQ",
+                "dateTime": datetime.now(tz=tzutc()).isoformat(),
+                "transactionId": str(uuid4()),
+            },
+            "payload": {
+                "fulfilmentRequest": {
+                    "email_address": self.email_address,
+                    "form_type": self.schema.form_type,
+                    "region_code": self.schema.region_code,
+                    "questionnaire_id": self.session_data.questionnaire_id,
+                    "tx_id": self.session_data.tx_id,
+                    "language_code": self.session_data.language_code,
+                    "display_address": self.session_data.display_address,
+                    "submitted_at": self.session_data.submitted_time,
+                }
+            },
+        }
```

Comment:
The `FulfilmentRequest` class should handle the top level of the message, as this is the same for both fulfilment requests. This could be done via a `message` property that calls this `payload` property:

```python
    def message(self) -> bytes:
        message = {
            "event": {
                "type": "FULFILMENT_REQUESTED",
                "source": "QUESTIONNAIRE_RUNNER",
                "channel": "EQ",
                "dateTime": datetime.now(tz=tzutc()).isoformat(),
                "transactionId": str(uuid4()),
            },
            "payload": self.payload
        }
        return json.dumps(message).encode("utf-8")
```
Then this method would become:

```python
    def message(self) -> Mapping:
        return {
            "fulfilmentRequest": {
                "email_address": self.email_address,
                "form_type": self.schema.form_type,
                "region_code": self.schema.region_code,
                "questionnaire_id": self.session_data.questionnaire_id,
                "tx_id": self.session_data.tx_id,
                "language_code": self.session_data.language_code,
                "display_address": self.session_data.display_address,
                "submitted_at": self.session_data.submitted_time,
            }
        },
     }
```

And the call to `fulfilment_request.payload` would be `fulfilment_request.message`.

Reply:
Good shout, have done with a slighl change. Made the payload a private method as it is not used by the client directly.